src/VisualStudio/Core/Def/Implementation/Library/FindResults/TreeItems/SourceReferenceTreeItem.cs

```diff
@@ -2,17 +2,67 @@
+using Microsoft.CodeAnalysis.Navigation;
 using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Extensions;
+using Roslyn.Utilities;
+using VsTextSpan = Microsoft.VisualStudio.TextManager.Interop.TextSpan;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.Library.FindResults
 {
-    internal class SourceReferenceTreeItem : AbstractSourceTreeItem, IComparable<SourceReferenceTreeItem>
+    internal class SourceReferenceTreeItem : AbstractTreeItem, IComparable<SourceReferenceTreeItem>
     {
+        protected readonly Workspace _workspace;
+        protected readonly DocumentId _documentId;
+        protected readonly string _projectName;
+        protected readonly string _filePath;
+        protected readonly TextSpan _sourceSpan;
+        protected readonly string _textLineString;
+        protected readonly int _lineNumber;
+        protected readonly int _offset;
+        protected readonly int _mappedLineNumber;
+        protected readonly int _mappedOffset;
+
+        private static readonly ObjectPool<StringBuilder> s_filePathBuilderPool = new ObjectPool<StringBuilder>(() => new StringBuilder());
+
+        private SourceReferenceTreeItem(
+            Document document, TextSpan sourceSpan, ushort glyphIndex, int commonPathElements)
+            : base(glyphIndex)
+        {
+            // We store the document ID, line and offset for navigation so that we
+            // still provide reasonable navigation if the user makes changes elsewhere
+            // in the document other than inserting or removing lines.
+
+            _workspace = document.Project.Solution.Workspace;
+            _documentId = document.Id;
+            _projectName = document.Project.Name;
+            _filePath = GetFilePath(document, commonPathElements);
+            _sourceSpan = sourceSpan;
+
+            var text = document.GetTextAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None);
+            var textLine = text.Lines.GetLineFromPosition(_sourceSpan.Start);
+            _textLineString = textLine.ToString();
+
+            _lineNumber = textLine.LineNumber;
+            _offset = sourceSpan.Start - textLine.Start;
+
+            var spanInSecondaryBuffer = text.GetVsTextSpanForLineOffset(_lineNumber, _offset);
+
+            VsTextSpan spanInPrimaryBuffer;
+            var succeeded = spanInSecondaryBuffer.TryMapSpanFromSecondaryBufferToPrimaryBuffer(_workspace, _documentId, out spanInPrimaryBuffer);
+
+            _mappedLineNumber = succeeded ? spanInPrimaryBuffer.iStartLine : _lineNumber;
+            _mappedOffset = succeeded ? spanInPrimaryBuffer.iStartIndex : _offset;
+        }
+
         public SourceReferenceTreeItem(
-            Document document, TextSpan sourceSpan, ushort glyphIndex, int commonPathElements,
-            string displayText = null, bool includeFileLocation = false)
-            : base(document, sourceSpan, glyphIndex, commonPathElements)
```

Comment:
instead of calling into the base, we just forward to the complicated constructor we copied in from AbstractSourceTreeItem.